### PR TITLE
"Entity identification" dimension, required utility methods and tests

### PR DESCRIPTION
**Pull request by ahemphill**

### CODE REVIEW

File: tools/scoring/score/dimensions/dimension.py [hunk 1]
```diff
@@ -113,8 +114,51 @@ def result_reporting(self) -> float:
...
+    Args:
+      entity: A standard entity instance
+
+    Returns:
+      Boolean indicating whether the entity has a `cloud_device_id`
+    """
+    return entity.cloud_device_id is not None
+
+  @staticmethod
+  def entity_is_virtual(*, entity: EntityInstance) -> bool:
```
Comment (charbull) [inline tools/scoring/score/dimensions/dimension.py:148]: nit: `is_virtual_entity` ?

Comment (charbull) [reply tools/scoring/score/dimensions/dimension.py:148]: just to be consistent with the other tooling on github

**Review by charbull — CHANGES_REQUESTED** [review]
nits and questions :)

File: tools/scoring/score/dimensions/dimension.py [hunk 2]
```diff
@@ -113,8 +114,51 @@ def result_reporting(self) -> float:
             self.correct_ceiling_reporting
             ) if self.correct_ceiling_reporting else None
 
+  @staticmethod
+  def entity_is_canonical(*, entity: EntityInstance) -> bool:
+    """ Utility for determining whether an entity is canonical.
+    Used in "complex" dimensions to filter sets for comparison.
+
+    Args:
+      entity: An entity instance which has been appended
```
[the remark this answers is not in the document]
Comment (ahemphill) [reply tools/scoring/score/dimensions/dimension.py:123]: Short answer: "I don't know" — but it's specified in the "Device Mapping Pilot Result Scoring" doc. It occurs to me now that the "simple" dimensions are not filtered this way presently or in the initial implementation; we should talk about whether/how they should be.

File: tools/scoring/score/dimensions/dimension.py [hunk 3]
```diff
@@ -113,8 +114,51 @@ def result_reporting(self) -> float:
             self.correct_ceiling_reporting
             ) if self.correct_ceiling_reporting else None
 
+  @staticmethod
+  def entity_is_canonical(*, entity: EntityInstance) -> bool:
```
[the remark this answers is not in the document]
Comment (ahemphill) [reply tools/scoring/score/dimensions/dimension.py:118]: Is that the preferred style? In terms of language, I wouldn't expect that to return anything appreciably different than as currently implemented (maybe "yes" or "no" versus "true" or "false"), but I prefer it as-is for two reasons: 1) the subject is the first word 2) the predicate matches the attribute being checked. Either way is fine.

File: tools/scoring/score/dimensions/dimension.py [hunk 4]
```diff
@@ -113,8 +114,51 @@ def result_reporting(self) -> float:
...
+  def entity_is_reporting(*, entity: EntityInstance) -> bool:
+    """
+    Utility for determining whether an entity is reporting.
+    Used in "complex" dimensions to filter sets for comparison.
+
+    Args:
+      entity: A standard entity instance
+
+    Returns:
+      Boolean indicating whether the entity has a `cloud_device_id`
```
[the remark this answers is not in the document]
Comment (ahemphill) [reply tools/scoring/score/dimensions/dimension.py:143]: Yes, however missing translations are counted in `ParseConfig.append_types()`. (This was a last-minute feature add and should be refactored as noted previously.)

File: tools/scoring/score/dimensions/dimension.py (hunk 1 above)
Comment (ahemphill) [reply tools/scoring/score/dimensions/dimension.py:148]: (As above)

File: tools/scoring/score/dimensions/entity_identification.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            self.entity_is_canonical,
+            filter(self.entity_is_reporting, solution.values()))
+    ]
+    proposed_reporting = [
+        entity.cloud_device_id for entity in filter(
+            self.entity_is_canonical,
+            filter(self.entity_is_reporting, proposed.values()))
+    ]
+
+    # Lists of `cloud_device_id`s representing
```
[the remark this answers is not in the document]
Comment (ahemphill) [reply tools/scoring/score/dimensions/entity_identification.py:48]: Clarified, I think

File: tools/scoring/score/dimensions/dimension.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahemphill) [reply tools/scoring/score/dimensions/dimension.py:123]: Added task.

**Review by charbull — APPROVED** [review]

**Pull request merged**